### PR TITLE
8287809: Revisit implementation of memory session

### PR DESCRIPTION
**Pull request by mcimadamore**
This is a cleanup of the memory session implementation. The main new concept is that `MemorySessionImpl` is split into two parts: there is an implementation of memory session and then there is a state abstraction (`MemorySessionImpl.State`). This allows to share the state across multiple session views, in a more clean way. The big consequence of this change is that the routines on `ScopedMemoryAccess` now have to be defined in terms of the state abstraction (but the changes are mostly mechanical).

I have consolidated the implementation quite a bit, by removing all the duplicated logic for issuing similar-looking exceptions. I have also addressed an issue with `checkValidState` throwing a _new_ `WrongThreadException` instead of using a singleton (which is what the logic for closing down shared session requires, to avoid stack walks that are too deep).

`MemorySession.State::checkValidState` is now fully monomorphic; when looking at benchmarks, this seems to be the best solution in order to make things fast. Specializing implmentations to remove few plain checks does not buy enough, and always has the risk of adding profile pollution.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287809](https://bugs.openjdk.org/browse/JDK-8287809): Revisit implementation of memory session

### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**) ⚠️ Review applies to [5b8f7246](https://git.openjdk.org/jdk/pull/9017/files/5b8f7246623b617edbbd178993bee14638194a1c)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9017/head:pull/9017` \
`$ git checkout pull/9017`

Update a local copy of the PR: \
`$ git checkout pull/9017` \
`$ git pull https://git.openjdk.org/jdk pull/9017/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9017`

View PR using the GUI difftool: \
`$ git pr show -t 9017`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9017.diff">https://git.openjdk.org/jdk/pull/9017.diff</a>

</details>

### CODE REVIEW

**Comment by mcimadamore** [comment]
Note: other cleanups are possible - such as flattening `ResourceList` into `MemorySession.State`. While that is possible, that's a tricky step, given that the session state can be registered against a cleaner. One option, which I explored, would be to use `MemorySessionImpl` for reachability - which then has an impact on `ScopedMemoryAccess`, as we would have to add reachability fences on the session, as well as on the session state. In any case, I'd like to proceed by steps, since this cleanup was already getting big enough.

File: src/java.base/share/classes/jdk/internal/foreign/SharedSessionState.java [hunk 1]
```diff
@@ -28,66 +28,69 @@
...
+ * A shared session state, which can be manipulated by multiple threads. Closing a shared state has to ensure that
+ * (i) only one thread can successfully close the state (e.g. in a close vs. close race) and that
+ * (ii) no other thread is accessing the memory associated with this state while the state is being
  * closed. To ensure the former condition, a CAS is performed on the liveness bit. Ensuring the latter
  * is trickier, and require a complex synchronization protocol (see {@link jdk.internal.misc.ScopedMemoryAccess}).
  * Since it is the responsibility of the closing thread to make sure that no concurrent access is possible,
  * checking the liveness bit upon access can be performed in plain mode, as in the confined case.
  */
-class SharedSession extends MemorySessionImpl {
+public class SharedSessionState extends MemorySessionImpl.State {
```
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/SharedSessionState.java:46]: Unclear to me why this is now public. It still seems to be only accessed from with the `jdk.internal.foreign` package?

File: src/java.base/share/classes/jdk/internal/foreign/ConfinedSessionState.java [hunk 2]
```diff
@@ -32,27 +32,27 @@
...
- * confinement check - that is, calling any operation on this session from a thread other than the
- * owner thread will result in an exception. Because of this restriction, checking the liveness bit
- * can be performed in plain mode.
+ * A confined session state, which features an owner thread. Because of this restriction, acquire and release
+ * can be implemented cheaply, with a plain field update. Closing the state is also cheap (since the liveness
+ * bit cannot be updated concurrently from other threads). Some extra complexity is required to support release
+ * operations triggered by threads other than the owner thread, which we support.
  */
-final class ConfinedSession extends MemorySessionImpl {
+public final class ConfinedSessionState extends MemorySessionImpl.State {
```
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/ConfinedSessionState.java:40]: Same here. I don't see why this is made public now.

Comment (JornVernee) [reply src/java.base/share/classes/jdk/internal/foreign/ConfinedSessionState.java:40]: Same for `justClose` and `ConfinedList::add` in this file. They still seem only accessed in the same package.

**Review by JornVernee — APPROVED** [review]
I think this looks good overall, but please try to limit the accessibility of  the methods in `State`.

**Comment by mcimadamore** [comment]
> I think this looks good overall, but please try to limit the accessibility of the methods in `State`.

Thanks, I'll take another look

**Pull request closed without merging**